src/transaction.c

```diff
@@ -358,36 +725,375 @@ int wally_tx_input_free(struct wally_tx_input *input)
-    if (!new_script)
+    unsigned char *new_script = src->script_len != 0 ? wally_malloc(src->script_len) : NULL;
+#ifdef BUILD_ELEMENTS
+    unsigned char *new_asset = src->asset_len != 0 ? wally_malloc(src->asset_len) : NULL;
+    unsigned char *new_value = src->value_len != 0 ? wally_malloc(src->value_len) : NULL;
+    unsigned char *new_nonce = src->nonce_len != 0 ? wally_malloc(src->nonce_len) : NULL;
+    unsigned char *new_surjectionproof = src->surjectionproof_len != 0 ?
+                                         wally_malloc(src->surjectionproof_len) : NULL;
+    unsigned char *new_rangeproof = src->rangeproof_len != 0 ?
+                                    wally_malloc(src->rangeproof_len) : NULL;
+#endif /* BUILD_ELEMENTS */
+    if ((src->script_len && !new_script)
+#ifdef BUILD_ELEMENTS
+        || (src->asset_len && !new_asset) ||
+           (src->value_len && !new_value) ||
+           (src->nonce_len && !new_nonce) ||
+           (src->surjectionproof_len && !new_surjectionproof) ||
+           (src->rangeproof_len && !new_rangeproof)
+#endif /* BUILD_ELEMENTS */
+    ) {
+#ifdef BUILD_ELEMENTS
+        wally_free(new_asset);
+        wally_free(new_value);
+        wally_free(new_nonce);
+        wally_free(new_surjectionproof);
+        wally_free(new_rangeproof);
+#endif /* BUILD_ELEMENTS */
         return false;
+    }
 
-    dst->satoshi = src->satoshi;
-    memcpy(new_script, src->script, src->script_len);
+    memcpy(dst, src, sizeof(*src));
+    if (src->script_len)
+        memcpy(new_script, src->script, src->script_len);
     dst->script = new_script;
-    dst->script_len = src->script_len;
+#ifdef BUILD_ELEMENTS
+    if (src->asset_len)
+        memcpy(new_asset, src->asset, src->asset_len);
+    if (src->value_len)
+        memcpy(new_value, src->value, src->value_len);
+    if (src->nonce_len)
+        memcpy(new_nonce, src->nonce, src->nonce_len);
+    if (src->surjectionproof_len)
+        memcpy(new_surjectionproof, src->surjectionproof, src->surjectionproof_len);
+    if (src->rangeproof_len)
+        memcpy(new_rangeproof, src->rangeproof, src->rangeproof_len);
+    dst->asset = new_asset;
+    dst->value = new_value;
+    dst->nonce = new_nonce;
+    dst->surjectionproof = new_surjectionproof;
+    dst->rangeproof = new_rangeproof;
+#endif /* BUILD_ELEMENTS */
     return true;
 }
 
-int wally_tx_output_init(uint64_t satoshi,
-                         const unsigned char *script, size_t script_len,
-                         struct wally_tx_output *output)
+static int tx_elements_output_proof_set(
+    struct wally_tx_output *output,
+    const unsigned char *surjectionproof,
+    size_t surjectionproof_len,
+    const unsigned char *rangeproof,
+    size_t rangeproof_len)
+{
+
+    if (((surjectionproof != NULL) != (surjectionproof_len != 0)) ||
+       ((rangeproof != NULL) != (rangeproof_len != 0)))
+        return WALLY_EINVAL;
+#ifdef BUILD_ELEMENTS
+    if (output->surjectionproof)
```

Comment:
nit: This check of the pointer before calling wally_free seems inconsistent with other places in the file